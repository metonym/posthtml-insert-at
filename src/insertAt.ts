import createMatcher from "posthtml-match-helper";
import { insertNode } from "./insertNode";
import type PostHTML from "posthtml";

type Options = IInsertAtData | IInsertAtData[];

export interface IInsertAtData {
  selector: string;
  behavior?: "inside" | "outside";
  append?: string;
  prepend?: string;
}

export function insertAt(options: Options) {
  return function plugin(tree: PostHTML.Node) {
    const opts = Array.isArray(options) ? options : [options];

    opts.forEach((option) => {
      const matcher = createMatcher(option.selector);
      const behavior = option.behavior || "inside";

      if (behavior === "inside") {
        tree.match(matcher, (node) =>
          insertNode({ node, option, content: [node.content] })
        );
      } else {
        let siblingNode = {};

        tree.match(matcher, (node) => {
          siblingNode = node;
          return node;
        });

        const matchingNode = siblingNode as PostHTML.Node["content"];

        tree.match({ content: [matcher] }, (node) =>
          insertNode({ node, option, content: [matchingNode] })
        );
      }
    });
  };
}
