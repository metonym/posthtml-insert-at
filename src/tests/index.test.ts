import posthtml from 'posthtml';
import insertAt from '../';
import { testCases } from './__fixtures__/testCases';

describe('insertAt', () => {
  testCases.forEach(testCase => {
    test(`'${testCase.name}' matches the snapshot`, () => {
      posthtml()
        .use(insertAt(testCase.options))
        .process(testCase.input)
        .then((result: { html: string }) => {
          expect(result.html).toMatchSnapshot();
        });
    });
  });
});
