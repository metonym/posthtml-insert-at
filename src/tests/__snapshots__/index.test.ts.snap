// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`insertAt 'Append/prepend inside' matches the snapshot 1`] = `
"
      <html>
        <body>
          <main><header></header>text<footer></footer></main>
        </body>
      </html>
    "
`;

exports[`insertAt 'Append/prepend outside' matches the snapshot 1`] = `
"
      <html>
        <body><header></header><main>text</main><footer></footer></body>
      </html>
    "
`;

exports[`insertAt 'Behavior not specified' matches the snapshot 1`] = `
"
      <html>
        <body>
          <main>text<footer></footer></main>
        </body>
      </html>
    "
`;

exports[`insertAt 'Options is an array' matches the snapshot 1`] = `
"
      <html>
        <body>
          <main>text<footer></footer></main>
          <div id=\\"content\\"><footer></footer></div>
        </body>
      </html>
    "
`;
